stratum/hal/lib/bcm/bcm_sdk_wrapper.cc

```diff
@@ -172,23 +865,633 @@ ::util::Status BcmSdkWrapper::InitializeSdk(
+
+  // Configure PORT_POLICY to classify packets with value 0x8100 at bytes 12,13
+  // as outer VLAN tagged packet.
+  uint64_t tmp[1] = {1};
+  RETURN_IF_BCM_ERROR(bcmlt_entry_allocate(unit, PORT_POLICYs, &entry_hdl));
+  for (std::vector<std::pair<int, int>> ::iterator it = tmp_ports.begin();
+       it != tmp_ports.end(); ++it) {
+    RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, PORT_IDs, it->first));
+    RETURN_IF_BCM_ERROR(
+        bcmlt_entry_field_array_add(entry_hdl, PASS_ON_OUTER_TPID_MATCHs, 0,
+                                    tmp, 1));
+    RETURN_IF_BCM_ERROR(
+        bcmlt_custom_entry_commit(entry_hdl, BCMLT_OPCODE_INSERT,
+                                  BCMLT_PRIORITY_NORMAL));
+  }
+  RETURN_IF_BCM_ERROR(bcmlt_entry_free(entry_hdl));
+
+  // Create L3_IIF_PROFILE 1 and enable IPv4 and IPv6 routing.
+  RETURN_IF_BCM_ERROR(bcmlt_entry_allocate(unit, L3_IIF_PROFILEs, &entry_hdl));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, L3_IIF_PROFILE_IDs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, IPV4_UCs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, IPV6_UCs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_custom_entry_commit(entry_hdl, BCMLT_OPCODE_INSERT,
+                                                BCMLT_PRIORITY_NORMAL));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_free(entry_hdl));
+
+  // Create L3_IIF index 1 and set VRF_ID=0.
+  RETURN_IF_BCM_ERROR(bcmlt_entry_allocate(unit, L3_IIFs, &entry_hdl));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, L3_IIF_IDs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, VRF_IDs, 0));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, L3_IIF_PROFILE_IDs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_custom_entry_commit(entry_hdl, BCMLT_OPCODE_INSERT,
+                                                BCMLT_PRIORITY_NORMAL));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_free(entry_hdl));
+
+  // Enable packet counters on all ports
+  // TODO(max): only add configured ports to bitmap, reduces polling CPU load
+  RETURN_IF_BCM_ERROR(bcmlt_entry_allocate(unit, CTR_CONTROLs, &entry_hdl));
+  auto _ = gtl::MakeCleanup([entry_hdl]() { bcmlt_entry_free(entry_hdl); });
+  RETURN_IF_BCM_ERROR(
+      bcmlt_entry_field_array_add(entry_hdl, PORTSs, 0, vlan_pbmp, 3));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, INTERVALs,
+                      FLAGS_port_counters_interval_in_usec));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, MULTIPLIER_PORTs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, MULTIPLIER_EPIPEs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, MULTIPLIER_IPIPEs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, MULTIPLIER_TMs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_entry_field_add(entry_hdl, MULTIPLIER_EVICTs, 1));
+  RETURN_IF_BCM_ERROR(bcmlt_custom_entry_commit(entry_hdl, BCMLT_OPCODE_UPDATE,
+                                                BCMLT_PRIORITY_NORMAL));
```

Comment:
I looked into the implementation of the lt entry handles and they're manually allocated and managed within one big chunk of memory. That's why missing entry_deletes never show up in ASAN builds. The big chunk is always free'd on shutdown.
I'll see if there is an easy way to get these leaks detected, then fixing the issues is a lot easier and reliable.

Reply:
I implemented basic handle leak detection in SDKLT and found no leaks in our code during a simple startup->shutdown sequence.
The cleanup running at block scope exit is not a problem, `[entry_handle]` captures the handle (a `uint32`) by value at closure creation, so reusing the same variable is possible.

Reply:
I see. This won't be a big issue since this only calls once. (only a few entries not be deallocated). Let me run some tests for `GetPortCounters` since it only frees the entry of error counter (I guess)

Reply:
I can see the memory keeps increase if I keep polling counter values via gNMI

Reply:
Actually, the memory won't be increased if I don't use the ASAN.
So this part looks good to me

Reply:
As show, the cleanup works. The could be more clear, but I don't see how at the moment.